### PR TITLE
Switch E3SM build system to CMake!

### PR DESCRIPTION
**Pull request by jgfouca**
This is the big one! This PR switches E3SM's default build-system to be cmake-based. This will be the most significant change to CIME since the pythonization of the case control system. Note that this change only impacts components, builds of sharedlibs will not change.

I had been attempting to integrate this work in smaller chunks, but I had to suspend these integrations in order not to disrupt things before the code freeze. This PR will also need to not be integrated until the code freeze is over.

At a high level, we are going from this:
before_cmake

To this:
after_cmake

Plot legend for above images:
* Squares = callstacks
* Ovals = input/output files
* Green = python
* Blue = perl
* Black = raw text
* Red = Make
* Purple = CMake

### Motivations
* Get the developer much "closer" to their builds. Instead of significant layers of CIME magic in-between the developer and their build, they can now simply cd to their builddir and type "ninja" (assuming .env_mach_specific.sh has been loaded).
* Reduce latency in development cycle. Due to being able to invoke ninja/make directly, all the overhead of case.build can be avoided, including all the namelist stuff.
* Reduce complexity/technical-debt of the existing system:
  - We have about 750 lines of CMake in components (compared to 1030 just in scripts/Tools/Makefile)
  - The old system had up to 4 process "layers" (python->perl->Make->perl), new has 2 (python->CMake, python->Ninja (auto-generated))
  - MPAS went from ~2k lines of Make to 640 lines of CMake
* Improve overall build times
  - Building components for test SMS_P12x2.ne4_oQU240.A_WCYCL1850.melvin_gnu.allactive-mach_mods went from 281 seconds to 178 seconds on melvin.
* The system no longer relies on any magic environment settings in CIME, so cmake/ninja calls should be trivial to replicate in a shell
* Bring E3SM closer in-line with other code bases
* Improve understanding of the E3SM build system
* Put E3SM on a more sustainable path going into the future. It will be much easier to integrate new codes (Kokkos) and components (SCREAM) into our build system.

### User impacts
* There is no longer separate build dirs, logs for each component. Since it all goes through a single CMake invocation, there's only one log file (e3sm.bldlog.XXX) and one build directory ($EXEROOT/bld). Each component will have it's own build subdirectory ($EXEROOT/bld/$compclass).
* All machines will need to have at least cmake-3.9 loaded in their config_machines.xml entry.
* If you decide to build your case outside of CIME (using Cmake and ninja directly), you'll still need to call case.build at the end of your dev-cycle in order to allow CIME to do various post-build bookkeeping.
* The CMake command used to configure the build will be the first thing in the e3sm.bldlog file. I've had success copy/pasting that command into a shell, allowing me to tinker with both CMake (and ninja) directly without having to go through CIME.
* Each component now has a build_cmake script in it's cime_config.
* You can go back to the "classic" build system by providing the --use-old options to case.build, this should work as long as the old buildlib scripts are maintained. Depending on how long CESM stays on the classic system, the cime/scripts/Tools/Makefile will likely be supported for quite a while.
* MPAS developers will need to keep their Make and CMake files in sync.
* We can now support various CMake backends. Ninja is our default backend since it's faster than make (especially for incremental builds) and its backend handles fortran dependencies correctly in the presence of preprocessor directives whereas the gmake does not always. You can force the gmake backed via --use-gmake to case.build.
* Component developers should never have to worry about fortran mod dependencies again, CMake takes care of that automatically (E3SM components were using a custom mkDepends script for this, MPAS was doing it by-hand in their Makefiles).

### XML changes
This PR disambiguates the selection of component specific flags in config_compiles.xml. We used to support the MODEL selector, IE:
```
<FFLAGS>
    <append MODEL="cam"> -qspill=6000 </append>
</FFLAGS>
```
But it was unclear whether MODEL referred to the component name ("cam") or component class ("atm"). The interpretation of MODEL was not even consistent across components. So MODEL is gone. Instead, please use COMP_NAME and COMP_CLASS.
```
<FFLAGS>
    <append COMP_NAME="cam"> -qspill=6000 </append>
</FFLAGS>
```

### Testing:
* As I switched each component over to the cmake system, I used the new bld_diff tool to confirm that the same files were being compiled with the same flags (with some expected exceptions). Now that cmake only has one log file, things are a bit harder to verify, but the flag logic is all the same as it was before this PR.
* With the current PR, e3sm_developer passes on melvin and is BFB.
* e3sm_integration builds on melvin (some fails, but they are on master too).

### Future work:
* The impact of the classic system is still felt in the new system and therefore our system still looks a bit unusual compared a "normal" code base. In particular, the interaction between the component/bld/configure perl scripts providing key inputs (Filepath, CCSM_cppdefs) to Cmake is a bit unusual. The jury is still out on the best path forward. The perl configure scripts are huge and CMake is not really a great general-purpose programming language, so it may not make sense to convert all this logic to CMake. @wlin7 is leading the effort to clean-up these scripts.
* The classic system also imposes outdated ways of thinking about a build system that will make it challenging to adopt more modern build system patterns. We are forced to think of things in terms of flags and directories. Modern CMake patterns encourage developers to think of things in terms of targets (each with public and private properties) and language features.
* It might be worth looking at the sharedlib builds to see if that can be cleaned-up as well

I would be super happy if others took the time to help me beta-test this. You'll need to be on this branch and set your mpas-source submodule to 20d6cdb0, which you can fetch from git@github.com:jgfouca/MPAS-Model.git.

In the meantime, please avoid (or consult with me) changes to the classic system.

@jedwards4b , @mvertens , @billsacks , @bmpersc 

[BFB]

### CODE REVIEW

**Comment by bishtgautam** [comment]
@jgfouca,  I noticed that lib for each component are now named by active/data/stub mode  (e.g. for I-compset `libdatam.a` is created instead of `libatm.a` a; and `libsice.a` is created instead of `libice.a`). Why is there a need for the for all components to share a single bld directory? Why can't each component use cmake and separate bld directories?

**Comment by jgfouca** [comment]
@bishtgautam,

> I noticed that lib for each component are now named by active/data/stub mode (e.g. for I-compset libdatam.a is created instead of libatm.a a; and libsice.a is created instead of libice.a)

Good catch. That wasn't an intentional change on my part. Is this going to cause problems?

> Why is there a need for the for all components to share a single bld directory?

Because we want developers to be able to build all of E3SM with a single cmake call, so we will definitely only have a single build directory. That said, it would probably be better if we had some subdirectories in our builds in order to avoid name conflicts in .mod files between components. If we were using CMake normally, we would have that. But due to the need to continue with the legacy $component/bld/configure system, I was not able to support components via Cmake's add_subdirectory.

**Comment by bishtgautam** [comment]
> Is this going to cause problems?

I caught this accidentally when I did the following
- First, `./case.build`
- Then, `./case.build --use-old` 
I had forgotten to call `./case.build --clean` in between the two commands. Since the two builds are creating libs that with different names, the previously generated libs aren't overwritten and at the linking stage all libs within the `lib/` dir are picked up. The creation of `e3sm.exe` fails as it complains about duplicate symbols within libs.

> it would probably be better if we had some subdirectories in our builds in order to avoid name conflicts in .mod files between components

That would have been very helpful to have. Maybe in the future ***if*** we decide to only support cmake, we can use add_subdirectory feature

A different question: Would it be possible for the makefile that is auto-generated by the Cmake to have an option to only compile individual models (i.e. `make libclm.a`)?

**Comment by jgfouca** [comment]
@bishtgautam , I definitely would not recommend mixing ./case.build --use-old and ./case.build in the same case. That's just begging to uncover untested corner cases.

> A different question: Would it be possible for the makefile that is auto-generated by the Cmake to have an option to only compile individual models (i.e. make libclm.a)?

That should be even easier than before:

```
make clm
```

**Comment by jgfouca** [comment]
I think there may be a way to hack add_subdirectory in.

**Comment by rljacob** [comment]
Renaming the libraries is fine with me.  That puts us on the road to allowing, for example, an F and I case in the same executable.  But we have to do a lot of other work for that to happen so you could change the names back for now.

**Comment by jgfouca** [comment]
@bishtgautam, I just realized a couple things:
1) The mpas libs were not being put in the correct directory. They were going in $case/lib instead of $case/bld/lib. Fix coming.
2) With CMake, there's no reason to put libs in bld/lib, so it may actually be viable to mix both classic and cmake builds in the same case.

**Comment by jgfouca** [comment]
OK, we now have separate build dirs for components. Also, note that I updated the MPAS sha that is needed.

File: cime/scripts/Tools/case.build [hunk 1]
```diff
@@ -62,6 +62,12 @@ def parse_command_line(args, description):
                         help="Case directory to build.\n"
                         "Default is current directory.")
 
+    parser.add_argument("--use-old", action="store_true",
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply cime/scripts/Tools/case.build]: You're right. --use-old only makes sense if model is e3sm.

**Comment by rljacob** [comment]
Does this override PR #2982 ?

**Comment by jgfouca** [comment]
@rljacob , yes.

**Comment by jgfouca** [comment]
@rljacob , I'd like to have a plan (with dates/timelines) for merging this. What needs to happen before this can go in?

**Comment by bishtgautam** [comment]
@jgfouca, Thanks for adding `make clm`. Few minor suggestions:
- Would it possible to not create empty directories (i.e. `bld/atm`, `bld/lnd`, ... `lib`)?
- Also, would it be possible to do the either of two following things?
  - Move the cmake stuff one directory up (i.e. move `bld/bld/*` into `bld/`), or
  - Rename `bld/bld`?

**Comment by jgfouca** [comment]
@bishtgautam ,

> Would it possible to not create empty directories

It might be possible someday. Unfortunately, those directories are needed if the user does './case.build --use-old' and there's no way to know at setup time if that might happen.

> Also, would it be possible to do the either of two following things?

Yes, definitely possible. If we use EXEROOT as the root cmake build directory, things might get a bit confusing since log files and exes would be getting mixed with cmake-generated files. I wasn't crazy about bld/bld either, can you think of a better name?

**Comment by bishtgautam** [comment]
> can you think of a better name?

How about `bld/cmake-bld`?

**Comment by bishtgautam** [comment]
One another thing that I have noticed on my mac laptop is that the build fails at the last step when linking the netcdf file. The `$(shell ${NETCDF_PATH}/bin/nf-config --flibs)` for SLIB from Macros.cmake shows up as `$(shell /opt/local/bin/nf-config --flibs)` in the e3sm.bldlog file. I'm using  cmake 3.13.4 on my laptop.

**Comment by jgfouca** [comment]
@bishtgautam , does that build on your mac laptop work when you use --use-old?

**Comment by bishtgautam** [comment]
e3sm does build with `./case.build --use-old` on my laptop

**Comment by rljacob** [comment]
Most of the v2.0alpha PRs need to be merged before this goes in.  Not possible to assign a date for that because it depends on testing/merging going smoothly.

**Comment by wlin7** [comment]
@jgfouca, just did a test to build a F-Case with COSP.

- First to add a note on an extra benefit with the new build system concerning external modules. It would check for the existence of those modules during setup stage and abort if not found. With the old build system, the process wouldn't be aborted until deep into the compilation.

- cmake has problem to build COSP.  1) In cmake/build_model.cmake, line 57, it tries to include Cosp.cmake from components/cmake/atm/cam/src/physics/cosp. The file is currently under component/cam/src/physics/cosp instead. 2) Even after resetting the include path, cmake appears to assume the source files in Cosp.cmake to have a path relative to "components/cmake/atm". The current Cosp.cmake has the the path relative to "components". I can make the model build by using "../../cam/src/physics/cosp/'  to let cmake recognize the relative path in build_model.cmake and Cosp.cmake. But I think you must have a more elegant solution.

**Comment by jgfouca** [comment]
@bishtgautam , I changed the cmake bld dir to $EXEROOT/cmake-bld.

I will investigate the NETCDF error.

**Comment by jgfouca** [comment]
@wlin7 , I must have broken that when I made different build directories for each component. Fix coming.

**Comment by jgfouca** [comment]
@wlin7 , fixed.

**Comment by wlin7** [comment]
Great, @jgfouca. It works very well now for my tests.

**Comment by wlin7** [comment]
@jgfouca , just found a new problem during testing of #3029. The problem may have been introduced by #2922, so reporting it here for now.

It is about using user mods. For example for EAM, codes under SourceMods/src.cam take precedence over standard codes with the same names under components/cam/src.

This is no more working. The build program near search for nor use the files under SourceMods/src.cam. Going back and check, at least since PR #2922 (via branch jgfouca/cam/cmakify), it has not been working. In that PR, although it still generated "Filepath" file that has SourceMods/src.cam at the top, the user mods under ourceMods/src.cam were not picked up to build the model.

The same problem may also happen to user mods for other components.

**Comment by mt5555** [comment]
In the first year of the E3SM project, we decided to not support the SourceMods capability, as the prefered way to test local source code changes being to use a git branch.    However, I believe CESM still supports SourceMods in CIME, and thus it has continued to work in E3SM.   Does E3SM test this feature?  and do we want to revive support for it?

**Comment by wlin7** [comment]
Thank @mt5555. I was not aware of that first year decision. Most likely there isn't a test for this feature, or it would have been picked up.  I am certain it has been working in E3SM until recently.  We may want to have a broader discussion if to revive it. Personally I find it very useful and have been using it all the time.

**Comment by worleyph** [comment]
SourceMods/ is still required to make case-specific changes to the MPAS streams files, isn't it? Does CMake have anything to do with this? (@philipwjones ?)

**Comment by philipwjones** [comment]
I believe that is still the case. Presumably the substitution of the replacement streams file comes after the build, so wouldn't think it would be affected by CMake build itself, but not sure how the related CIME script changes might impact this.

**Comment by rljacob** [comment]
> We have about 750 lines of CMake in components
Is that the total over all the active components?

**Comment by jgfouca** [comment]
@wlin7 , thanks, I will look at this. If the classic build system still supported sourcemods, then the new one probably should as well.

**Comment by jgfouca** [comment]
> Is that the total over all the active components?

@rljacob , yes, over E3SM components (not including MPAS).

**Comment by wlin7** [comment]
That would be great, @jgfouca, thanks. However, I lost track when this feature stopped working. The development repo that I used extensively with the SourceMods feature is the master up to Jan 19, 2018. (ce9be2a24d82a9fe61db48e5687553a15bd2d8c8).

**Comment by billsacks** [comment]
@jgfouca for what it's worth, CISM - the one CESM science component that uses a CMake build - supports SourceMods, via a mechanism put in place by @agsalin : https://github.com/ESCOMP/cism/blob/master/CMakeLists.txt

**Comment by PeterCaldwell** [comment]
The reason why sourcemods was deprecated in the first year of E3SM is that sourcemods makes provenance very difficult. If all changes are on a branch, it is very easy to see what was actually run. If random code was injected into the build process, it is almost impossible to tell what's going on. I'm not sure who gets to make the call about whether sourcemods is supported, but I would vote for purposefully disabling it going forward rather than trying to fix this capability for cmake.

**Comment by wlin7** [comment]
@PeterCaldwell , thanks for refreshing our minds concerning SourceMods. It is true it would make provenance more difficult. The build system always creates a tar ball for the SourceMods, this is probably designed as a remedy (but wouldn't be as well preserved overtime as an all-inclusive branch). That said, the benefit of having SourceMods feature is also clear. 

@jgfouca , FYI, this feature is not supported in master even before merging PR #2922. It works in maint-1.0.

**Comment by jgfouca** [comment]
@wlin7 , @PeterCaldwell , @rljacob 

OK, maybe we shouldn't support SourceMods. With the build-system being much more streamlined with this PR, and Git being very capable of managing file changes, I don't know if the SourceMod workflow makes much sense. It just seems so much more natural to just modify the files in-place.

For what it's worth, it looks to me like SourceMods should still work with this PR. For example, the cam configure script still has this code:
```
    if ($ccsm_seq) {
        my $CASEROOT = "$ENV{'CASEROOT'}";
        print $fh "$CASEROOT/SourceMods/src.cam\n";
    }
```
Which should add the SourceMods dir to the top of the Filepath file, which gets used in both the classic and cmake systems.

**Comment by jgfouca** [comment]
I can officially remove support for SourceMods for E3SM or we can keep it limping along as a deprecated feature, let me know.

**Comment by jonbob** [comment]
The MPAS components do use the SourceMods directory to support changes to the streams files, and that usage is critical to us.

On Jul 16, 2019, at 10:57 AM, James Foucar <notifications@github.com<mailto:notifications@github.com>> wrote:


@wlin7<https://github.com/wlin7> , @PeterCaldwell<https://github.com/PeterCaldwell> , @rljacob<https://github.com/rljacob>

OK, maybe we shouldn't support SourceMods. With the build-system being much more streamlined with this PR, and Git being very capable of managing file changes, I don't know if the SourceMod workflow makes much sense. It just seems so much more natural to just modify the files in-place.

For what it's worth, it looks to me like SourceMods should still work with this PR. For example, the cam configure script still has this code:

    if ($ccsm_seq) {
        my $CASEROOT = "$ENV{'CASEROOT'}";
        print $fh "$CASEROOT/SourceMods/src.cam\n";
    }


Which should add the SourceMods dir to the top of the Filepath file, which gets used in both the classic and cmake systems.

**Comment by jgfouca** [comment]
@jonbob , ok, we will keep it.

**Comment by rljacob** [comment]
We'll keep it as a deprecated feature.  We'll have to eventually find a new way to make changes to stream files.

**Comment by wlin7** [comment]
@jgfouca, although SourceMods is always listed at the top  in Filepath, the actual build either does not search for files under the SourceMods path or not let such files take precedence. The model would still be built using the files under standard path even when there are modified user mods present.  

If you would do a test, you may copy ~wlin/ACME/integration/dev/dcape-ull/new/*.F90 to SoureceMods/src.cam of an existing casedir, rebuild the model and check if the user mods take  precedence - I would check in e3sm.bldlog for the path to file like physpkg.F90. Thanks.

**Comment by billsacks** [comment]
Added some notes from today's cime telecon here: https://github.com/esmci/cime/wiki/Meeting-Notes#cmake-build

**Comment by jgfouca** [comment]
@wlin7, as far as I can tell, the SourceMods feature is working.

My test:
```
% ./create_test SMS_Ln5.ne4_ne4.FC5AV1C-L --no-build
% cd $casedir/SourceMods/src.cam
% cp $srcroot/components/cam/src/chemistry/modal_aero/modal_aero_gasaerexch.F90 .
* inject syntax error into modal_aero_gasaerexch.F90 *
% cd ../..
% ./case.build
* observe that a build error for model_areo.. is triggered *
```

**Comment by wlin7** [comment]
That is interesting, @jgfouca. Supposedly copying any EAM source codes to SourceMods/src.cam would have no difference concerning testing this feature. I will do further tests.

**Comment by wlin7** [comment]
@jgfouca, it still does not work for me. I am using the branch for this PR ( 2c66234f2050b04facb35646a4f2c83e96efbd66). 

Moreover, some strange thing happened when I did a test just like you did

- create_test SMS_Ln5.ne4_ne4.FC5AV1C-L --no-build --machine=cori-knl

then copy the same file over and inserted a line of text that would fail the compiler. The build did fail. However, the build would also fail even when no invalid line was inserted, or no file was copied to SourceMods/src.cam. In my test, the failure apparently was due to other reason. Can you please check if the same thing happened in your test (say, repeating case.build after removing the file from SourceMods/src.cam)? Thanks.

**Comment by jgfouca** [comment]
@wlin7 

I tested:
1. No mods: worked
2. Copied a working file into srcmods: worked
3. Injected a syntax error into this srcmod file: worked (build failed with expected error on expected file)

**Comment by wlin7** [comment]
Thanks for the update, @jgfouca. Are you using branch jgfouca/cime/case_build_use_old? I wish to reproduce your success then figure why my test didn't work out. I used both create_test, and regular create_case.

**Comment by jgfouca** [comment]
@wlin7 , yes I'm just using the HEAD of this PR branch.

**Comment by vanroekel** [comment]
@jgfouca this branch does not work for me on summit.  It fails quickly in the mpas build.  It is a vanilla build with pgi.  No special changes or flags.  It looks like the NETCDF_PATH variable is not getting to the MPAS build.  But I have no idea how to fix this.  Could you please try this on summit and let me know if this work for you?  I've reset the MPAS source branch to the tag you sent me.  Here is my create case command.
```
./create_newcase -case testbuild -compset GMPAS-IAF -mach summit -res T62_oEC60to30v3 -compiler pgi -proj cli115
```

**Comment by jgfouca** [comment]
@vanroekel , OK I will look at it.

**Comment by jgfouca** [comment]
@vanroekel , I can't even build an A case on summit on either this branch or master. It fails building csm_share:

```
"/autofs/nccs-svm1_home1/acmetest/ACME-clean/cime/src/share/util/shr_flux_mod.F90", line 2454.26: 1513-041 (S) Arguments of the wrong type were specified for the INTRINSIC procedure "min".
"/autofs/nccs-svm1_home1/acmetest/ACME-clean/cime/src/share/util/shr_flux_mod.F90", line 2481.27: 1513-041 (S) Arguments of the wrong type were specified for the INTRINSIC procedure "max".
"/autofs/nccs-svm1_home1/acmetest/ACME-clean/cime/src/share/util/shr_flux_mod.F90", line 2532.16: 1513-041 (S) Arguments of the wrong type were specified for the INTRINSIC procedure "min".
"/autofs/nccs-svm1_home1/acmetest/ACME-clean/cime/src/share/util/shr_flux_mod.F90", line 2573.16: 1513-041 (S) Arguments of the wrong type were specified for the INTRINSIC procedure "min".
```

**Comment by jedwards4b** [comment]
@jgfouca it looks like constants in those lines need to be promoted - 
instead of .1 in line 2481 it should be .1_R8

**Comment by jgfouca** [comment]
Thanks, @jedwards4b . Turns out i was using the wrong compiler (intel instead of pgi).

**Comment by jgfouca** [comment]
@vanroekel , I was able to get SMS.T62_oEC60to30v3.GMPAS-IAFt to build with some changes to this PR. If you pull this branch and set MPAS to 20d6cdb0, it should work.

**Comment by vanroekel** [comment]
thanks @jgfouca! It's working now.

**Comment by jgfouca** [comment]
TODO items: enjoy consistent use of MODEL selector in XML files. Make ninja the default backend.

**Comment by jgfouca** [comment]
Did another upstream merge and quick re-test on melvin. All looks good! Merging to next.

**Pull request merged**